src/main/java/org/openlmis/buq/service/buq/BottomUpQuantificationService.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * This program is part of the OpenLMIS logistics management information system platform software.
+ * Copyright © 2017 VillageReach
+ *
+ * This program is free software: you can redistribute it and/or modify it under the terms
+ * of the GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details. You should have received a copy of
+ * the GNU Affero General Public License along with this program. If not, see
+ * http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
+ */
+
+package org.openlmis.buq.service.buq;
+
+import static org.openlmis.buq.i18n.MessageKeys.ERROR_BOTTOM_UP_QUANTIFICATION_NOT_FOUND;
+import static org.openlmis.buq.i18n.MessageKeys.ERROR_FACILITY_NOT_FOUND;
+import static org.openlmis.buq.i18n.MessageKeys.ERROR_ID_MISMATCH;
+import static org.openlmis.buq.i18n.MessageKeys.ERROR_PROCESSING_PERIOD_NOT_FOUND;
+import static org.openlmis.buq.i18n.MessageKeys.ERROR_PROGRAM_NOT_FOUND;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.openlmis.buq.domain.buq.BottomUpQuantification;
+import org.openlmis.buq.domain.buq.BottomUpQuantificationLineItem;
+import org.openlmis.buq.domain.buq.BottomUpQuantificationStatus;
+import org.openlmis.buq.domain.buq.BottomUpQuantificationStatusChange;
+import org.openlmis.buq.dto.buq.BottomUpQuantificationDto;
+import org.openlmis.buq.dto.referencedata.ApprovedProductDto;
+import org.openlmis.buq.dto.referencedata.FacilityDto;
+import org.openlmis.buq.dto.referencedata.ProcessingPeriodDto;
+import org.openlmis.buq.dto.referencedata.ProgramDto;
+import org.openlmis.buq.exception.ContentNotFoundMessageException;
+import org.openlmis.buq.exception.ValidationMessageException;
+import org.openlmis.buq.i18n.MessageKeys;
+import org.openlmis.buq.repository.buq.BottomUpQuantificationRepository;
+import org.openlmis.buq.service.referencedata.ApprovedProductReferenceDataService;
+import org.openlmis.buq.service.referencedata.FacilityReferenceDataService;
+import org.openlmis.buq.service.referencedata.PeriodReferenceDataService;
+import org.openlmis.buq.service.referencedata.ProgramReferenceDataService;
+import org.openlmis.buq.util.AuthenticationHelper;
+import org.openlmis.buq.util.FacilitySupportsProgramHelper;
+import org.openlmis.buq.util.Message;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@SuppressWarnings("PMD.TooManyMethods")
+@Service
```

Comment:
shouldn't we add `@Transactional` annotation here?

Reply:
I stuck to the OpenLMIS convention (@Transactional in controllers, none in services)

Reply:
Marked as resolve per slack discussion.